From d2237c19da1077984dfbf9fecec759e503717622 Mon Sep 17 00:00:00 2001
From: JiangWu <wujiang@kylinos.cn>
Date: Thu, 19 May 2022 15:47:14 +0800
Subject: [PATCH] randr: Correctly get physical size for screen with RandR 1.5

According to the existing calculation method, multiple monitors
are spliced into a virtual monitor, and there are some problems
in the calculation of the virtual physical size. The existing
way to calculate the virtual physical size: the resolution of
the last monitor is divided by the resolution of the first monitor,
and then multiplied by the physical size of the first monitor.

For example, two monitors are spliced into a virtual monitor,
namely VGA-0(1920/477x1080/268+0+0)), DVI-0(1680/477x1050/268+1920+0).
Because the resolution of DVI-0 is smaller than that of VGA-0,
the virtual physical size will be 0 (1680/1920*477=0, 1050/1080*268=0).

The virtual physical size should be the virtual display resolution
divided by the first display resolution multiplied by the first display
physical size. For example mmWidth=double(3600/1920)*477, mmHeight=
double(1080/1080)*268.

Signed-off-by: Jiang Wu <wujiang@kylinos.cn>
---
 randr/rrmonitor.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/randr/rrmonitor.c b/randr/rrmonitor.c
index e62bd484d..7b0bbfcd5 100644
--- a/randr/rrmonitor.c
+++ b/randr/rrmonitor.c
@@ -156,8 +156,8 @@ RRMonitorGetGeometry(RRMonitorPtr monitor, RRMonitorGeometryPtr geometry)
 
         /* Adjust physical sizes to account for total area */
         if (active_crtcs > 1 && first.box.x2 != first.box.x1 && first.box.y2 != first.box.y1) {
-            geometry->mmWidth = (this.box.x2 - this.box.x1) / (first.box.x2 - first.box.x1) * first.mmWidth;
-            geometry->mmHeight = (this.box.y2 - this.box.y1) / (first.box.y2 - first.box.y1) * first.mmHeight;
+            geometry->mmWidth = ((double)(geometry->box.x2 - geometry->box.x1) / (first.box.x2 - first.box.x1)) * first.mmWidth;
+            geometry->mmHeight = ((double)(geometry->box.y2 - geometry->box.y1) / (first.box.y2 - first.box.y1)) * first.mmHeight;
         }
     } else {
         *geometry = monitor->geometry;
-- 
GitLab

